### PR TITLE
Training

### PR DESCRIPTION
**Pull request by samwang141224**
This new branch includes codes for training the CNN model. The codes for training may be difficult to test, so I just describe how to do the training in the notebook.

### CODE REVIEW

File: desidlas/training/model.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
+
+""" TensorFlow Models for DLA finder"""
+''' new model with smooth training, this method can improve the DLA claasification on low signal-to-noise '''
```
Comment (profxj) [inline desidlas/training/model.py]: typo:  "classification"

File: desidlas/training/model.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+tf.nn.conv2d(input, filter, strides, padding, use_cudnn_on_gpu=None, name=None)
+"""
+import tensorflow as tf 
+import math
+
+def weight_variable(shape):
+    
+    """
+    Generating random number according to the tensor shape,The standard deviation is 0.1，define a variable function
+    parameter_shape:the shape of the output tensor，A 1-D integer Tensor or Python array
```
Comment (profxj) [inline desidlas/training/model.py]: do you just mean `shape` ?

File: desidlas/training/model.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+    """
+    initial = tf.random.truncated_normal(shape, stddev=0.1) #Outputs random values from a truncated normal distribution，shape:The shape of the output tensor.stddev:The standard deviation of the normal distribution, before truncation
+    return tf.Variable(initial)
+
+
+def bias_variable(shape):
+    
+    """
+    Generating constant 0 according to the tensor shape,define a variable function
+    parameter_shape:the shape of the output tensor，A 1-D integer Tensor or Python array
```
Comment (profxj) [inline desidlas/training/model.py]: Same as above.

Comment (profxj) [reply desidlas/training/model.py]: Oh, I see this is a different doc style.
I'm not familiar with it.  Might be best
to stick with Google or numpy style.

File: desidlas/training/model.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    
+    
+    """
+    This is Model_v5 from SDSS
+    
+    three conv layers,three pool layers,two full connected layers
+    Parameters
+    ----------
+    hyperparameters：use the hyperparameters in dla_cnn/models/model_gensample_v7.1_hyperparams.json
+    INPUT_SIZE:pixels numbers for each window , 400 for high SNR and 600 for low SNR
```
Comment (profxj) [inline desidlas/training/model.py]: I think there needs to be a space after the :

File: desidlas/training/model.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+    pool1_stride = hyperparameters['pool1_stride']
+    pool2_stride = hyperparameters['pool2_stride']
+    pool3_stride = hyperparameters['pool3_stride']
+    pool1_method = 1        # Removed from hyperparameter search because it never chose anything but this method
+    pool2_method = 1
+    pool3_method = 1
+
+    #INPUT_SIZE = 600
+    #matrix_size=4
+     #hyperparameters['INPUT_SIZE'] # the pixel of input image,need a new hyperparameter
```
Comment (profxj) [inline desidlas/training/model.py]: remove this commented text?

File: desidlas/training/model.py [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+     #matrix_size:smoothed data input, change the input training data to 600*4 scale . without smoothing , the matrix_size is 1
+    tfo = {}    # Tensorflow objects
+
+
+    
+
+    """
+
+    x = tf.Variable(tf.ones(shape=[None, INPUT_SIZE]), name='x')
```
Comment (profxj) [inline desidlas/training/model.py]: Add some inline comments.

Comment (profxj) [reply desidlas/training/model.py]: Or is this code commented out?  If so, remove?

File: desidlas/training/model.py [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+    """
+
+    # I converted "tf.placeholder" to "tf.Variable"
+
+    tf.compat.v1.disable_eager_execution()
+
+    x = tf.compat.v1.placeholder(tf.float32, shape=[None,matrix_size, INPUT_SIZE], name='x')
+    y = tf.compat.v1.placeholder(tf.float32, shape=[None, matrix_size], name='x')
+
+    label_classifier = tf.compat.v1.placeholder(tf.float32, shape=[None], name='label_classifier')
```
Comment (profxj) [inline desidlas/training/model.py:180]: Add inline comments.

File: desidlas/training/model.py [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+
+     
+
+ 
+
+        
+
+def build_model(hyperparameters,INPUT_SIZE,matrix_size):
+    
+
```
Comment (profxj) [inline desidlas/training/model.py:120]: It would be good somewhere in the Repo to describe the standard model.  
Could be here, or could be in a separate doc.

But please provide it.  It should include discussion on stride, kernel size, loss function, optimizer, etc.

File: desidlas/training/model.py [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+                    tf.compat.v1.div(label_coldensity,label_coldensity+epsilon)) +
+        l2_regularization_penalty * (tf.nn.l2_loss(W_conv1) + tf.nn.l2_loss(W_conv2) +
+                                     tf.nn.l2_loss(W_fc1) + tf.nn.l2_loss(W_fc2_1)),
+        name='loss_coldensity_regression')
+
+    optimizer = tf.compat.v1.train.AdamOptimizer(learning_rate=learning_rate)
+
+    cost_all_samples_lossfns_AB = loss_classifier + loss_offset_regression
+    cost_pos_samples_lossfns_ABC = loss_classifier + loss_offset_regression + loss_coldensity_regression
+    # train_step_AB = tf.train.AdamOptimizer(learning_rate=learning_rate).minimize(cost_all_samples_lossfns_AB, global_step=global_step, name='train_step_AB')
```
Comment (profxj) [inline desidlas/training/model.py]: Remove or make as an option?

File: desidlas/training/model.py [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+    keep_prob = tf.compat.v1.placeholder(tf.float32, name='keep_prob')
+    global_step = tf.Variable(0, name='global_step', trainable=False)
+
+    x_4d = tf.reshape(x, [-1, INPUT_SIZE, 1,matrix_size]) #reshape the data size
+    # First Convolutional Layer
+    # Kernel size (16,1)
+    # Stride (4,1)
+    # number of filters = 4 (features?)
+    # Neuron activation = ReLU (rectified linear unit)
+    W_conv1 = weight_variable([conv1_kernel, 1, 4, conv1_filters])
```
Comment (profxj) [inline desidlas/training/model.py:192]: Can we make the `4` a parameter?

File: desidlas/training/model.py [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+    # Train and Evaluate the model
+    loss_classifier = tf.add(tf.nn.sigmoid_cross_entropy_with_logits(logits=y_nn_classifier, labels=label_classifier),
+                             l2_regularization_penalty * (tf.nn.l2_loss(W_conv1) + tf.nn.l2_loss(W_conv2) +
+                                                          tf.nn.l2_loss(W_fc1) + tf.nn.l2_loss(W_fc2_1)),
+                             name='loss_classifier')
+    loss_offset_regression = tf.add(tf.reduce_sum(input_tensor=tf.nn.l2_loss(y_nn_offset - label_offset)),
+                                    l2_regularization_penalty * (tf.nn.l2_loss(W_conv1) + tf.nn.l2_loss(W_conv2) +
+                                                                 tf.nn.l2_loss(W_fc1) + tf.nn.l2_loss(W_fc2_2)),
+                                    name='loss_offset_regression')
+    epsilon = 1e-6
```
Comment (profxj) [inline desidlas/training/model.py]: This should probably be a parameter in the method call.

File: desidlas/training/model.py [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+    train_step_ABC = optimizer.minimize(cost_pos_samples_lossfns_ABC, global_step=global_step, name='train_step_ABC')
+    # train_step_C = tf.train.AdamOptimizer(learning_rate=learning_rate).minimize(loss_coldensity_regression, global_step=global_step, name='train_step_C')
+    output_classifier = tf.sigmoid(y_nn_classifier, name='output_classifier')
+    prediction = tf.round(output_classifier, name='prediction')
+    correct_prediction = tf.equal(prediction, label_classifier)
+    accuracy = tf.reduce_mean(input_tensor=tf.cast(correct_prediction, tf.float32), name='accuracy')
+    rmse_offset = tf.sqrt(tf.reduce_mean(input_tensor=tf.square(tf.subtract(y_nn_offset,label_offset))), name='rmse_offset')
+    rmse_coldensity = tf.sqrt(tf.reduce_mean(input_tensor=tf.square(tf.subtract(y_nn_coldensity,label_coldensity))), name='rmse_coldensity')
+
+    variable_summaries(loss_classifier, 'loss_classifier', 'SUMMARY_A')
```
Comment (profxj) [inline desidlas/training/model.py:315]: What do these lines do?  Please add a comment.

File: desidlas/training/training.py [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+
+from tensorflow.python.framework import ops
+ops.reset_default_graph()
+
+from model import build_model
+
+from tensorflow.compat.v1 import ConfigProto
+from tensorflow.compat.v1 import InteractiveSession
+
+config = ConfigProto()
```
Comment (profxj) [inline desidlas/training/training.py:24]: What are these lines doing?  Add a comment.

File: desidlas/training/training.py [hunk 14]
```diff
@@ -0,0 +1,397 @@
...
+config.gpu_options.allow_growth = True
+config.allow_soft_placement=True
+config.log_device_placement = True
+
+
+tensor_regex = re.compile('.*:\d*')
+
+# Get a tensor by name, convenience method
+def t(tensor_name):
```
Comment (profxj) [inline desidlas/training/training.py:33]: Add doc string

File: desidlas/training/training.py [hunk 15]
```diff
@@ -0,0 +1,397 @@
...
+    #inputs: label_classifier, label_offset, label_coldensity
+    """
+    Perform training and estimate accuracy,RMSE after every iteration training
+    
+    Parameters
+    ----------
+    sess:tf.Session object
+    ixs:np.ndarray, indices for each window in the dataset
+    data:sightline data in training and testing dataset
+
```
Comment (profxj) [inline desidlas/training/training.py:59]: Explain `summary_writer`

File: desidlas/training/training.py [hunk 16]
```diff
@@ -0,0 +1,397 @@
...
+    for pos_ix in pos_ixs_split:
+        tensors = [t('accuracy'), t('loss_classifier'), t('rmse_offset'), t('loss_offset_regression'),
+                        t('rmse_coldensity'), t('loss_coldensity_regression'), t('global_step')]
+        if summary_writer is not None:
+            tensors.extend(tf.compat.v1.get_collection('SUMMARY_A'))
+            tensors.extend(tf.compat.v1.get_collection('SUMMARY_B'))
+            tensors.extend(tf.compat.v1.get_collection('SUMMARY_C'))
+            tensors.extend(tf.compat.v1.get_collection('SUMMARY_shared'))
+
+        run = sess.run(tensors, feed_dict={t('x'):                data['fluxes'][pos_ix],
```
Comment (profxj) [inline desidlas/training/training.py:98]: Add inline comments

File: desidlas/training/training.py [hunk 17]
```diff
@@ -0,0 +1,397 @@
...
+                          i, train_accuracy, float(np.mean(loss_value)), result_rmse_offset,
+                             result_loss_offset_regression, result_rmse_coldensity,
+                             result_loss_coldensity_regression))
+                    record_file.write("step {:06d}, classify-offset-density acc/loss - RMSE/loss      {:0.3f}/{:0.3f} - {:0.3f}/{:0.3f} - {:0.3f}/{:0.3f}\n".format(
+                          i, train_accuracy, float(np.mean(loss_value)), result_rmse_offset,
+                             result_loss_offset_regression, result_rmse_coldensity,
+                             result_loss_coldensity_regression))
+                    
+                # Run on test
+                if i % 5000 == 0 or i == training_iters - 1:
```
Comment (profxj) [inline desidlas/training/training.py:222]: Should 5000 be a parameter?

File: desidlas/training/training.py [hunk 18]
```diff
@@ -0,0 +1,397 @@
...
+    best_coldensity_rmse:the minimum column density during the training
+
+    Returns:
+    -------
+    normalized_score:used to select hyperparameters. A better hyperparameter combine will have a smaller normalized_score
+
+    """
+    # These mean & std dev are used to normalize the score from all 3 loss functions for hyperparam optimize
+    # TODO -- Reconsider these for DESI
+    mean_best_accuracy = 0.02
```
Comment (profxj) [inline desidlas/training/training.py:280]: Were these optimized?  If not, should they be?

Are these reported in the paper?

File: desidlas/training/training.py [hunk 19]
```diff
@@ -0,0 +1,397 @@
...
+    std_best_coldensity_rmse = 0.1
+    normalized_score = (1 - best_accuracy - mean_best_accuracy) / std_best_accuracy + \
+                       (best_offset_rmse - mean_best_offset_rmse) / std_best_offset_rmse + \
+                       (best_coldensity_rmse - mean_best_coldensity_rmse) / std_best_coldensity_rmse
+    
+    return normalized_score
+
+
+
+if __name__ == '__main__':
```
Comment (profxj) [inline desidlas/training/training.py:294]: The code that follows contains a lot of hard coded info, e.g. paths.

Please make this a separate script.

File: desidlas/training/training.py [hunk 20]
```diff
@@ -0,0 +1,397 @@
...
+    INPUT_SIZE = args['INPUT_SIZE']
+    matrix_size = args['matrix_size']
+
+    train_dataset = Dataset(args['train_dataset_filename'])
+    test_dataset = Dataset(args['test_dataset_filename'])
+
+    exception_counter = 0
+    iteration_num = 0
+
+    parameter_names = ["learning_rate", "training_iters", "batch_size", "l2_regularization_penalty", "dropout_keep_prob",
```
Comment (profxj) [inline desidlas/training/training.py]: Is this defining the model?

If so, make it its own file with a sensible versioning schema.

**Review by profxj — CHANGES_REQUESTED** [review]
Great work.  Please consider the suggested comments.
Thanks.

File: desidlas/training/model.py (hunk 1 above)
Comment (samwang141224) [reply desidlas/training/model.py]: sorry, I have corrected it

File: desidlas/training/model.py (hunk 2 above)
Comment (samwang141224) [reply desidlas/training/model.py]: Thanks for the comments, I have changed the 'parameter_shape' to 'shape' and make the comments for every module have the same style

File: desidlas/training/model.py (hunk 4 above)
Comment (samwang141224) [reply desidlas/training/model.py]: Thanks, I added it.

File: desidlas/training/model.py (hunk 5 above)
Comment (samwang141224) [reply desidlas/training/model.py]: done

File: desidlas/training/model.py (hunk 6 above)
Comment (samwang141224) [reply desidlas/training/model.py]: Sorry, I forgot to remove them.

File: desidlas/training/model.py (hunk 8 above)
Comment (samwang141224) [reply desidlas/training/model.py:120]: Do you mean that we should describe more details of this model in the paper? We just show the basic structure of the CNN and show the model parameters in the paper.

File: desidlas/training/model.py (hunk 9 above)
Comment (samwang141224) [reply desidlas/training/model.py]: I just remove them.

File: desidlas/training/model.py (hunk 10 above)
Comment (samwang141224) [reply desidlas/training/model.py:192]: I think 4 may not be a variable parameter. I just make the variable parameter when we do the smoothing as matrix_size.

File: desidlas/training/model.py (hunk 12 above)
Comment (samwang141224) [reply desidlas/training/model.py:315]: done

File: desidlas/training/model.py (hunk 7 above)
Comment (samwang141224) [reply desidlas/training/model.py:180]: done

File: desidlas/training/model.py (hunk 11 above)
Comment (samwang141224) [reply desidlas/training/model.py]: This value is used in the loss function of column density, Parks chose this value because he said this is a small value safe from 32-bit floating point rounding error.

File: desidlas/training/model.py (hunk 12 above)
Comment (samwang141224) [reply desidlas/training/model.py:315]: Hi X, I just have another small question. When I reply to your comments, I just reply one by one. This will send you many emails (one reply one email) or just one email that includes all my replies? If it is the first case, I think this may disturb you a lot. So if it is the first case, could you tell me how to do a better replying? If it is the second case, then I will just keep replying like before. Thanks!

File: desidlas/training/training.py (hunk 13 above)
Comment (samwang141224) [reply desidlas/training/training.py:24]: These lines are used for parameter configuration for tensorflow session, I have added inline comments.

File: desidlas/training/training.py (hunk 14 above)
Comment (samwang141224) [reply desidlas/training/training.py:33]: done

File: desidlas/training/training.py (hunk 15 above)
Comment (samwang141224) [reply desidlas/training/training.py:59]: done

File: desidlas/training/training.py (hunk 17 above)
Comment (samwang141224) [reply desidlas/training/training.py:222]: I think this may not be necessary to set this as a parameter. This code will print the testing accuracy every 5000 steps, I think this value is fine for the training?

File: desidlas/training/training.py (hunk 18 above)
Comment (samwang141224) [reply desidlas/training/training.py:280]: They are not optimized, just as same as what Parks chose. We do not describe these values in detail in the paper, we just report we do the hyperparameter search.

File: desidlas/training/training.py (hunk 20 above)
Comment (samwang141224) [reply desidlas/training/training.py]: I have put them in a  new file: parameterset.py

File: desidlas/training/training.py (hunk 19 above)
Comment (samwang141224) [reply desidlas/training/training.py:294]: Thanks. I have replaced the hard coded info.

File: desidlas/training/model.py (hunk 8 above)
Comment (samwang141224) [reply desidlas/training/model.py:120]: I have made a file describing the CNN model, data input, loss function, and outputs in desidlas/training/CNN model.md

File: desidlas/training/model.py (hunk 12 above)
Comment (profxj) [reply desidlas/training/model.py:315]: I like them one by one.  Thx.

**Review by profxj — APPROVED** [review]
great stuff!
